src/internal/components/LayoutEditor.tsx

```diff
@@ -0,0 +1,234 @@
+              {
+                id: layoutSaveState.hash,
+                state: jsonFromEscapedJsonString(layoutSaveState.history),
+              },
+            ]
+          : [
+              {
+                id: layoutSaveState.hash,
+                state: jsonFromEscapedJsonString(layoutSaveState.history),
+              },
+            ],
+        index: visualConfigurationData ? 1 : 0,
+        appendData: false,
+      });
+      setPuckInitialHistoryFetched(true);
+      return;
+    }
+
+    const localHistoryIndex = JSON.parse(localHistoryArray).findIndex(
+      (item: any) => item.id === layoutSaveState?.hash
+    );
+
+    // If local storage reset Puck history to it
+    if (localHistoryIndex !== -1) {
+      devLogger.log(
+        "Layout Prod Mode - Using localStorage visual configuration"
+      );
+      setPuckInitialHistory({
+        histories: JSON.parse(localHistoryArray),
+        index: localHistoryIndex,
+        appendData: false,
+      });
+      setPuckInitialHistoryFetched(true);
+      return;
+    }
+
+    // otherwise start fresh - this user doesn't have localStorage that reflects the saved state
+    clearVisualConfigLocalStorage();
+  }, [
+    setPuckInitialHistory,
+    setPuckInitialHistoryFetched,
+    clearVisualConfigLocalStorage,
+    buildVisualConfigLocalStorageKey,
+  ]);
+
+  useEffect(() => {
+    loadPuckInitialHistory();
+  }, [templateMetadata]);
+
+  // Log PUCK_INITIAL_HISTORY (layout) on load
+  useEffect(() => {
+    if (puckInitialHistory) {
+      devLogger.logData("PUCK_INITIAL_HISTORY", puckInitialHistory);
+    }
+  }, [puckInitialHistory]);
+
+  // Handles sending the devSaveStateData on initial load so that nothing isn't rendered for preview.
+  // Subsequent changes are sent via handleHistoryChange in InternalLayoutEditor.tsx
+  useEffect(() => {
+    if (!puckInitialHistoryFetched) {
+      return;
+    }
+
+    const historyToSend =
+      puckInitialHistory?.histories[puckInitialHistory.histories.length - 1]
+        .state.data;
+
+    devLogger.logFunc("sendDevSaveStateData useEffect");
+    sendDevLayoutSaveStateData({
+      payload: { devSaveStateData: JSON.stringify(historyToSend) },
+    });
+  }, [puckInitialHistoryFetched, puckInitialHistory, templateMetadata]);
+
+  const isLoading = !puckInitialHistoryFetched;
+
+  return (
+    <>
+      {!isLoading ? (
+        <InternalLayoutEditor
+          puckConfig={puckConfig}
+          puckInitialHistory={puckInitialHistory}
+          isLoading={isLoading}
+          clearHistory={clearHistory}
+          templateMetadata={templateMetadata}
+          layoutSaveState={layoutSaveState}
+          saveLayoutSaveState={saveLayoutSaveState}
+          publishVisualConfiguration={publishVisualConfiguration}
+          sendDevSaveStateData={sendDevLayoutSaveStateData}
+          buildVisualConfigLocalStorageKey={buildVisualConfigLocalStorageKey}
+        />
+      ) : (
+        <LoadingScreen progress={90} />
```

Comment:
Is LoadingScreen progress={90} purposeful?

Reply:
I set Editor.tsx loads to load from 0 to 90 and then ThemeEditor/LayoutEditor will either be at 90 or be loaded. I'm not sure that's the best way to do it

Reply:
Maybe annoying, but in Editor.tsx instead of showing loading there at all, pass the number of loading things through which can be added here and show the loading state in each sub editor. Then you can add to that number however many other things need to load in this component to show progress properly.

Reply:
Tried that a bit but the problem is we still have to wait for templateMetadata before selecting a sub editor, so we still need a loading in the top level editor.

What I did in the latest commit is just have Editor do the 0-100 loading with no loading bar in the subeditors. The subeditors are just waiting for PuckInitialHistory and/or ThemeInitialHistory, both of which do not make any network requests so they load very quickly. https://dev.yext.com/s/1000152098/yextsites/63009/editor#view=EDIT&entityId=8725521 seems just as smooth as it is in production right now.